Code review request

zdrale opened the pull request:
-In this project, we have implemented our own version of Ruby’s enumerable methods.
-These are the methods we recreated:
my_each
my_each_with_index
my_select
my_all?
my_any?
my_none?
my_count
my_map
my_inject
multiply_els

zdrale committed:
Updated project folder

-Added linters
-Added ruby file

zdrale committed:
Added my_each method

zdrale committed:
Added my_each_with_index method

zdrale committed:
Added my_select method

Naqhid committed:
did my_all, my_any, my_none

zdrale committed:
Removed unnecessary indentation

zdrale committed:
Added my_count method

Naqhid committed:
my_map

zdrale committed:
Added my_inject and multiply_els method

zdrale committed:
Update for linters

--Fixed 42 issues

zdrale committed:
Update for linters

--Fixed identation errors

zdrale committed:
Update for linters

--Fixed indentation

zdrale committed:
Updated methods file

--Aligned end statements with if

zdrale committed:
Removed whitespaces

arslanbisharat commented on methods.rb:
```diff
@@ -0,0 +1,112 @@
+    for i in self
+      yield i
+    end
+  end
```
* [ ] should have working `my_each`  identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_each
- [ ] my_each returns an Enumerator if no block is given

**- Yours is always expecting a block**
<img width="440" alt="3" src="https://user-images.githubusercontent.com/58504655/98254870-9cbf2b00-1f9e-11eb-92db-c74067e22c7d.PNG">


**- Real each**
![image](https://user-images.githubusercontent.com/52209067/77675424-8a820680-6f52-11ea-9013-f537927bd221.png)

arslanbisharat commented on methods.rb:
```diff
@@ -0,0 +1,112 @@
+      yield item, i
+      i += 1
+    end
+  end
```
* [ ] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_each_with_index

- [ ] my_each_with_index returns an Enumerator if no block is given

- Yours is always expecting a block
<img width="459" alt="4" src="https://user-images.githubusercontent.com/58504655/98255237-02abb280-1f9f-11eb-8998-89623a14020b.PNG">


- Real each_with_index
![image](https://user-images.githubusercontent.com/52209067/77680236-4ba37f00-6f59-11ea-8423-28dfc3dbf205.png)

arslanbisharat commented on methods.rb:
```diff
@@ -0,0 +1,112 @@
+      arry << i if yield i
+    end
+    arry
+  end
```
* [ ] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_select
- [ ] my_select returns an enumerator if no block is given

**- Yours is always expecting a block**
<img width="526" alt="5" src="https://user-images.githubusercontent.com/58504655/98255397-2ff86080-1f9f-11eb-92bf-c70077936122.PNG">


**- Real select**
![image](https://user-images.githubusercontent.com/52209067/77681510-0bdd9700-6f5b-11ea-9340-ac116b7ed62e.png)

arslanbisharat commented on methods.rb:
```diff
@@ -0,0 +1,112 @@
+      end
+    end
+    sum
+  end
```
- [ ] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```

- Real inject
![image](https://user-images.githubusercontent.com/52209067/77698409-2ffaa180-6f76-11ea-91d3-9778aa560411.png)

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is always expecting a block
<img width="438" alt="11" src="https://user-images.githubusercontent.com/58504655/98257055-0dffdd80-1fa1-11eb-873a-1282e8db26f5.PNG">


**- Real inject**
![image](https://user-images.githubusercontent.com/52209067/77698534-6a643e80-6f76-11ea-9d0e-ebb0f02da179.png)

arslanbisharat commented on methods.rb:
```diff
@@ -0,0 +1,112 @@
+    count = 0
+    count += 1 if size > 1
+    yield count
+  end
```
* [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_count

- [ ] my_count returns the number of items in enum through enumeration
```ruby
[5, 4, 1].my_count == [5, 4, 1].count => 3
```
- Yours is always expecting a block
<img width="434" alt="9" src="https://user-images.githubusercontent.com/58504655/98256566-85813d00-1fa0-11eb-9681-d457b6bf7b7d.PNG">


- Real count
![image](https://user-images.githubusercontent.com/52209067/77696292-c200ab00-6f72-11ea-8e97-5a2b79256cbc.png)

- [ ] my_count counts the number of items in enum that are equal to the item if an argument is given
```ruby
[5, 4, 1].my_count(4) == [5, 4, 1].count(4) => 1
```
- Yours is giving an ArgumentError
<img width="426" alt="10" src="https://user-images.githubusercontent.com/58504655/98256645-a184de80-1fa0-11eb-85bb-6e6a81f42872.PNG">

- Real count
![image](https://user-images.githubusercontent.com/52209067/77696378-e2306a00-6f72-11ea-9310-0bb30ab9c94d.png)

arslanbisharat reviewed (changes requested):
# Request changes 😳

I am very happy to see this project and I can see you worked hard on this project. There are only a few issues that we need to address and I am sure that you will end up with a perfect project

_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

**Please have a look at the suggestions to get more understanding.**

- [ ] Please remove linter erros. 

- [x] Please add a proper description in your readme file.

- [x] You have 10 failing tests. I have given suggestions below. Please have a look at them and solve them. A potential source for learning is [this](https://www.rubyguides.com/2016/03/enumerable-methods/)

**Please make the changes and request re-review for the given project**
Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

zdrale committed:
Update README.md

zdrale committed:
Update README.md

zdrale committed:
Updated methods

--Added check for if block is given
--Updated my_count to work as original
--Updated my_inject to work as original

zdrale committed:
Merge branch 'methods' of https://github.com/zdrale/Enumerable-Methods into methods

mimipeshy commented:
## STATUS :CHANGES REQUIRED :red_circle: 

Hey guys :wave:  :wave: , great job on working on most of the feedback given by the previous TSE, however we still have a few more changes to work on :smile: 

- [ ] Work on the linter errors you are having


Check the comments for more details on the same

Regards
Peris Ndanu

mimipeshy commented on methods.rb:
```diff
@@ -0,0 +1,123 @@
+    output
+  end
+
+  def my_count (argument=nil)
```
- [ ] If a block is given, it counts the number of elements yielding a true value.
![true](https://user-images.githubusercontent.com/33067256/98527585-43037d00-228c-11eb-8867-7c7d4bc6aae0.png)

- What it expects `array.count{ |x| x%2==0 } #=> 3`

- [ ] it should not change the original array
![mutate](https://user-images.githubusercontent.com/33067256/98527431-10f21b00-228c-11eb-93f7-dec07af93c75.png)

mimipeshy replied:
#my_count
- [ ] If a block is given, it counts the number of elements yielding a true value.
![true](https://user-images.githubusercontent.com/33067256/98527585-43037d00-228c-11eb-8867-7c7d4bc6aae0.png)

- What it expects `array.count{ |x| x%2==0 } #=> 3`

- [ ] it should not change the original array
![mutate](https://user-images.githubusercontent.com/33067256/98527431-10f21b00-228c-11eb-93f7-dec07af93c75.png)

mimipeshy commented on methods.rb:
```diff
@@ -0,0 +1,123 @@
+    new_arr
+  end
+
+  def my_inject(init = nil, arg = nil)
```
- [ ] enumerables #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
![inject](https://user-images.githubusercontent.com/33067256/98528093-e2287480-228c-11eb-83b9-338d4350c8ce.png)

- [ ]  enumerables #my_inject when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method: range

![range](https://user-images.githubusercontent.com/33067256/98528400-49debf80-228d-11eb-8404-5c1a3c4ac6e7.png)

mimipeshy commented on methods.rb:
```diff
@@ -0,0 +1,123 @@
+    sum
+  end
+
+  def multiply_els(arr)
```
- [ ] enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
![els](https://user-images.githubusercontent.com/33067256/98528723-bfe32680-228d-11eb-839b-fb2f6a717e4e.png)

mimipeshy commented on methods.rb:
```diff
@@ -0,0 +1,123 @@
+    counter = to_a.length
+  end
+  counter
+  end  
```
#my_count
- [ ] If a block is given, it counts the number of elements yielding a true value.
![true](https://user-images.githubusercontent.com/33067256/98527585-43037d00-228c-11eb-8867-7c7d4bc6aae0.png)

- What it expects `array.count{ |x| x%2==0 } #=> 3`

- [ ] it should not change the original array
![mutate](https://user-images.githubusercontent.com/33067256/98527431-10f21b00-228c-11eb-93f7-dec07af93c75.png)

mimipeshy commented on methods.rb:
```diff
@@ -0,0 +1,123 @@
+      end
+    end
+    sum
+  end
```
- [ ] enumerables #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method

**Yours is always expecting a block**
![inject](https://user-images.githubusercontent.com/33067256/98528093-e2287480-228c-11eb-83b9-338d4350c8ce.png)


- [ ]  enumerables #my_inject when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method: range

**Yours is always expecting a block**
![range](https://user-images.githubusercontent.com/33067256/98528400-49debf80-228d-11eb-8404-5c1a3c4ac6e7.png)

mimipeshy commented on methods.rb:
```diff
@@ -0,0 +1,123 @@
+    arr.my_inject { |n, total| n * total }
+  end
+
+end
```
- [ ] enumerables #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
![els](https://user-images.githubusercontent.com/33067256/98528723-bfe32680-228d-11eb-839b-fb2f6a717e4e.png)

Naqhid committed:
rectify errors

layzrd commented on methods.rb:
```diff
@@ -0,0 +1,128 @@
+    sum
+  end
+
+  def multiply_els(arr)
```
Move the #multiply_els method out of the module.

layzrd commented on methods.rb:
```diff
@@ -0,0 +1,128 @@
+    new_arr
+  end
+
+  def my_inject(init = nil, arg = nil)
```
#my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range.
Fix the method to make it work for a symbol.

layzrd commented on methods.rb:
```diff
@@ -0,0 +1,128 @@
+  end
+
+  def my_count (argument=nil)
+    if num
```
undefined local variable or method `num'. fix it.

layzrd reviewed (changes requested):
### Status: Require changes  ♲
Hi @zdrale  👏 You did very well on this project. ✨ I applaud for successfully implementing custom enumerable methods. Most requirements have done 👌, Kindly fix the following issues.

[REQUIRED]

- [ ] Fix syntax  errors that I comment inline the code
- [ ] Fix the my_inject method to enable it to accept symbols. You have to add all features to make your methods work like the original ruby enumerator. Check the following link https://ruby-doc.org/core-2.7.2/Enumerable.html.
- [ ] Fix the rubocop linting errors.

Once you fix the issues, kindly submit another code review. ♲

Happy coding 👨‍💻
Abdulaziz Redi Ali

Naqhid committed:
fix review errors

Naqhid committed:
fix linter errors

Naqhid committed:
update-code

Naqhid committed:
fix errors

zdrale committed:
Updated my_inject method

--Method can take symbol as a parameter now

zdrale committed:
Merge branch 'methods' of https://github.com/zdrale/Enumerable-Methods into methods

zdrale committed:
Update for linters

-Fixed linters errors

zdrale committed:
Deleted duplicate method

zdrale committed:
Added necessary whitespaces

zdrale committed:
Update for linters

-Fixed parentheses erros

jubaan reviewed (approved):
# :tada: :tada: :tada: :rotating_light: Status: Approved :rotating_light:  :tada: :tada: :tada:

#### :clap: :clap: Great job :hammer_and_pick:!! There are no more issues left to be addressed. 

- [x] Please fix all linter errors before merging to master.

## See you at the top :rocket:!

# :tada: :tada: :tada: :rotating_light: Status: Approved :rotating_light:  :tada: :tada: :tada:
